File: android/src/main/java/com/apollohg/editor/NativeProseViewerExpoView.kt
package com.apollohg.editor

import android.content.Context
import android.view.View
import android.view.ViewGroup
import expo.modules.kotlin.AppContext
import expo.modules.kotlin.viewevent.EventDispatcher
import expo.modules.kotlin.views.ExpoView

/** Expo adapter over the public Android prose viewer. */
class NativeProseViewerExpoView(
    context: Context,
    appContext: AppContext
) : ExpoView(context, appContext), ProseViewerInteractionListener {
    private val viewer = ProseViewerView(context)
    private val onContentHeightChange by EventDispatcher<Map<String, Any>>()
    @Suppress("unused")
    private val onPressLink by EventDispatcher<Map<String, Any>>()
    @Suppress("unused")
    private val onPressMention by EventDispatcher<Map<String, Any>>()

    private var lastRenderJson: String? = null
    private var lastThemeJson: String? = null
    private var lastEmittedContentHeight = 0
    internal var suppressContentHeightEventsForTesting = false
    internal val viewerForTesting: ProseViewerView
        get() = viewer

    init {
        importantForAccessibility = View.IMPORTANT_FOR_ACCESSIBILITY_NO
        viewer.interactionListener = this
        viewer.opensLinksAutomatically = true
        viewer.setCollapsesWhenEmpty(true)
        viewer.onContentHeightChange = { heightPx ->
            val heightWithHostPadding = if (viewer.isContentCollapsedForHost) {
                0
            } else {
                heightPx + paddingTop + paddingBottom
            }
            emitContentHeightIfNeeded(heightWithHostPadding, force = true)
        }
        addView(
            viewer,
            LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT
            )
        )
    }

    fun setRenderJson(renderJson: String?) {
        if (lastRenderJson == renderJson) return
        lastRenderJson = renderJson
        applyToViewer()
    }

    fun setThemeJson(themeJson: String?) {
        if (lastThemeJson == themeJson) return
        lastThemeJson = themeJson
        applyToViewer()
    }

    fun setImageLoadingPolicyJson(policyJson: String?) {
        viewer.setImageLoadingPolicyJson(policyJson)
    }

    fun setCollapsesWhenEmpty(collapsesWhenEmpty: Boolean?) {
        viewer.setCollapsesWhenEmpty(collapsesWhenEmpty ?: true)
    }

    fun setEnableLinkTaps(enableLinkTaps: Boolean?) {
        viewer.linkTapsEnabled = enableLinkTaps ?: true
    }

    fun setInterceptLinkTaps(interceptLinkTaps: Boolean?) {
        viewer.opensLinksAutomatically = !(interceptLinkTaps ?: false)
    }

    private fun applyToViewer() {
        viewer.apply(
            renderJson = lastRenderJson ?: "[]",
            themeJson = lastThemeJson ?: "{}"
        )
        requestLayout()
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        val childWidthSpec = getChildMeasureSpec(
            widthMeasureSpec,
            paddingLeft + paddingRight,
            viewer.layoutParams.width
        )
        viewer.measure(
            childWidthSpec,
            View.MeasureSpec.makeMeasureSpec(0, View.MeasureSpec.UNSPECIFIED)
        )
        val desiredWidth = viewer.measuredWidth + paddingLeft + paddingRight
        val desiredHeight = if (viewer.isContentCollapsedForHost) {
            0
        } else {
            viewer.measuredHeight + paddingTop + paddingBottom
        }
        setMeasuredDimension(
            resolveSize(desiredWidth, widthMeasureSpec),
            desiredHeight
        )
        emitContentHeightIfNeeded(desiredHeight)
    }

    override fun onLayout(
        changed: Boolean,
        left: Int,
        top: Int,
        right: Int,
        bottom: Int
    ) {
        val childTop = paddingTop
        viewer.layout(
            paddingLeft,
            childTop,
            right - left - paddingRight,
            childTop + viewer.measuredHeight
        )
        emitContentHeightIfNeeded(
            if (viewer.isContentCollapsedForHost) 0 else measuredHeight
        )
    }

    override fun onLinkTap(view: ProseViewerView, href: String, text: String) {
        onPressLink(mapOf("href" to href, "text" to text))
    }

    override fun onMentionTap(view: ProseViewerView, docPos: Long, label: String) {
        onPressMention(mapOf("docPos" to docPos, "label" to label))
    }

    private fun emitContentHeightIfNeeded(contentHeight: Int, force: Boolean = false) {
        if (contentHeight <= 0 && !viewer.isContentCollapsedForHost) return
        if (!force && contentHeight == lastEmittedContentHeight) return
        lastEmittedContentHeight = contentHeight
        if (!suppressContentHeightEventsForTesting) {
            onContentHeightChange(mapOf("contentHeight" to contentHeight))
        }
    }
}

File: ios/NativeProseViewerExpoView.swift
import ExpoModulesCore
import UIKit

/// Expo adapter over the public UIKit prose viewer.
final class NativeProseViewerExpoView: ExpoView {
    let onContentHeightChange = EventDispatcher()
    let onPressLink = EventDispatcher()
    let onPressMention = EventDispatcher()

    private let viewer = ProseViewerView(frame: .zero)
    private var lastRenderJSON: String?
    private var lastThemeJSON: String?
    private var lastEmittedContentHeight: CGFloat = 0
    private var lastMeasuredWidth: CGFloat = 0
    private var collapsesWhenEmpty = true

    internal var viewerForTesting: ProseViewerView { viewer }

    required init(appContext: AppContext? = nil) {
        super.init(appContext: appContext)
        viewer.interactionDelegate = self
        viewer.opensLinksAutomatically = true
        viewer.setCollapsesWhenEmpty(true)
        viewer.contentInset = UIEdgeInsets(top: 8, left: 0, bottom: 8, right: 0)
        viewer.onContentHeightChange = { [weak self] measuredHeight in
            self?.emitContentHeightIfNeeded(measuredHeight: measuredHeight, force: true)
        }
        addSubview(viewer)
    }

    var imageLoadingPolicy: ImageLoadingPolicy {
        viewer.imageLoadingPolicyForHost
    }

    func setImageLoadingPolicyJson(_ json: String?) {
        viewer.setImageLoadingPolicy(json: json)
    }

    func setEnableLinkTaps(_ enabled: Bool?) {
        viewer.linkTapsEnabled = enabled ?? true
    }

    func setInterceptLinkTaps(_ intercept: Bool?) {
        viewer.opensLinksAutomatically = !(intercept ?? false)
    }

    func setCollapsesWhenEmpty(_ collapses: Bool?) {
        let nextValue = collapses ?? true
        guard collapsesWhenEmpty != nextValue else { return }
        collapsesWhenEmpty = nextValue
        viewer.setCollapsesWhenEmpty(nextValue)
        setNeedsLayout()
        emitContentHeightIfNeeded(force: true)
    }

    func setRenderJson(_ renderJson: String?) {
        guard lastRenderJSON != renderJson else { return }
        lastRenderJSON = renderJson
        applyToViewer()
    }

    func setThemeJson(_ themeJson: String?) {
        guard lastThemeJSON != themeJson else { return }
        lastThemeJSON = themeJson
        let theme = EditorTheme.from(json: themeJson)
        let cornerRadius = theme?.borderRadius ?? 0
        layer.cornerRadius = cornerRadius
        clipsToBounds = cornerRadius > 0
        applyToViewer()
    }

    private func applyToViewer() {
        viewer.apply(
            renderJson: lastRenderJSON ?? "[]",
            themeJson: lastThemeJSON ?? "{}"
        )
        lastMeasuredWidth = 0
        invalidateIntrinsicContentSize()
        setNeedsLayout()
    }

    override var intrinsicContentSize: CGSize {
        if viewer.isContentCollapsedForHost {
            return CGSize(width: UIView.noIntrinsicMetric, height: 0)
        }
        guard lastEmittedContentHeight > 0 else {
            return CGSize(
                width: UIView.noIntrinsicMetric,
                height: UIView.noIntrinsicMetric
            )
        }
        return CGSize(width: UIView.noIntrinsicMetric, height: lastEmittedContentHeight)
    }

    override func layoutSubviews() {
        super.layoutSubviews()
        viewer.frame = bounds

        let currentWidth = ceil(bounds.width)
        guard abs(currentWidth - lastMeasuredWidth) > 0.5 else { return }
        lastMeasuredWidth = currentWidth
        emitContentHeightIfNeeded(force: true)
    }

    private func emitContentHeightIfNeeded(
        measuredHeight: CGFloat? = nil,
        force: Bool = false
    ) {
        let contentHeight: CGFloat
        if viewer.isContentCollapsedForHost {
            contentHeight = 0
        } else {
            guard bounds.width > 0 else { return }
            let fittedHeight = measuredHeight
                ?? viewer.measuredHeight(forWidth: bounds.width)
            contentHeight = ceil(fittedHeight)
            guard contentHeight > 0 else { return }
        }
        guard force || abs(contentHeight - lastEmittedContentHeight) > 0.5 else {
            return
        }
        lastEmittedContentHeight = contentHeight
        invalidateIntrinsicContentSize()
        onContentHeightChange(["contentHeight": contentHeight])
    }
}

extension NativeProseViewerExpoView: ProseViewerInteractionDelegate {
    func proseViewer(_ view: ProseViewerView, didTapLink href: String, text: String) {
        onPressLink(["href": href, "text": text])
    }

    func proseViewer(
        _ view: ProseViewerView,
        didTapMention docPos: Int,
        label: String
    ) {
        onPressMention(["docPos": docPos, "label": label])
    }
}
